### PR TITLE
Integrate proof check on sync

### PR DESCRIPTION
**Pull request by ssd04**
## Reasoning behind the pull request
- Wait for confirmation block in epochstart bootstrap
- integrate equivalent proofs into process sync
- use real proofs pool component in interceptors

## Testing procedure
- Will be tested with the feat branch

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: process/factory/interceptorscontainer/shardInterceptorsContainerFactory_test.go [hunk 1]
```diff
@@ -90,6 +91,10 @@ func createShardDataPools() dataRetriever.PoolsHolder {
 	pools.CurrBlockTxsCalled = func() dataRetriever.TransactionCacher {
 		return &mock.TxForCurrentBlockStub{}
 	}
+	pools.ProofsCalled = func() dataRetriever.ProofsPool {
+		return &dataRetrieverMock.ProofsPoolMock{}
```
Comment (sstanculeanu) [inline process/factory/interceptorscontainer/shardInterceptorsContainerFactory_test.go:95]: the mocked component should return by default this mocked proofs pool, so this would be avoided

File: process/interceptors/processor/hdrInterceptorProcessor.go [hunk 2]
```diff
@@ -68,6 +74,12 @@ func (hip *HdrInterceptorProcessor) Save(data process.InterceptedData, _ core.Pe
 
 	hip.headers.AddHeader(interceptedHdr.Hash(), interceptedHdr.HeaderHandler())
 
+	// TODO: check for equivalent flag
```
Comment (sstanculeanu) [inline process/interceptors/processor/hdrInterceptorProcessor.go:77]: :+1:

File: process/sync/baseSync.go [hunk 3]
```diff
@@ -687,6 +698,48 @@ func (boot *baseBootstrap) syncBlock() error {
...
+
+	// process block only if there is a proof for it
+	hasProof := boot.proofs.HasProof(header.GetShardID(), headerHash)
+	if hasProof {
+		return nil
+	}
+
+	log.Trace("baseBootstrap.handleEquivalentProof: did not have proof for header, will try again", "headerHash", headerHash)
+
+	// TODO: evaluate adding a wait here, or request for next header if missing
```
Comment (sstanculeanu) [inline process/sync/baseSync.go:733]: not sure this would work as expected, to be seen in tests

File: process/factory/interceptorscontainer/shardInterceptorsContainerFactory_test.go (hunk 1 above)
Comment (ssd04) [reply process/factory/interceptorscontainer/shardInterceptorsContainerFactory_test.go:95]: "called" fields are not set in this mocked component, so we have to set them here

File: process/interceptors/processor/hdrInterceptorProcessor.go (hunk 2 above)
Comment (ssd04) [reply process/interceptors/processor/hdrInterceptorProcessor.go]: added flag check

**Review by sstanculeanu — DISMISSED** [review]

File: process/sync/baseSync.go (hunk 3 above)
Comment (ssd04) [reply process/sync/baseSync.go]: the integration test is working in this case, will see with more testing if there are other edge cases here

File: epochStart/bootstrap/epochStartMetaBlockProcessor.go [hunk 4]
```diff
@@ -180,27 +239,90 @@ func (e *epochStartMetaBlockProcessor) GetEpochStartMetaBlock(ctx context.Contex
...
+	}
+
+	err = e.waitForConfMetaBlock(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	chanRequests := time.After(durationBetweenReRequests)
 	chanCheckMaps := time.After(durationBetweenChecks)
+
```
Comment (AdoAdoAdo) [inline epochStart/bootstrap/epochStartMetaBlockProcessor.go:253]: from the original process loop now we have three, but not sure we need 3 process loops.
Can you check if we can simplify?

With minimum changes from the previous version I would see the following flow:
- previously there was a request for the epoch start metablock and decide (statistically) which is the correct one. We were using the same process loop for both the request and validation that we have a "consensus" on the epoch start block.

- now in addition to the epoch start meta block, we also need its confirmation block (to take the proof for the previous block). To do this we don't need to decide statistically if there is a consensus on that block as well, it just needs to fit on top of the decided epoch start meta block.

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

File: dataRetriever/dataPool/proofsCache/proofsPool.go [hunk 5]
```diff
@@ -58,9 +62,20 @@ func (pp *proofsPool) AddProof(
 
 	proofsPerShard.addProof(headerProof)
 
+	pp.callAddedProofHandlers(headerProof)
+
 	return nil
 }
 
+func (pp *proofsPool) callAddedProofHandlers(headerProof data.HeaderProofHandler) {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply dataRetriever/dataPool/proofsCache/proofsPool.go:71]: done

File: epochStart/bootstrap/epochStartMetaBlockProcessor.go (hunk 4 above)
Comment (ssd04) [reply epochStart/bootstrap/epochStartMetaBlockProcessor.go]: updated to remove separated flow for confirmation block

File: process/sync/metablock_test.go [hunk 6]
```diff
@@ -1812,3 +1831,197 @@ func TestMetaBootstrap_SyncAccountsDBs(t *testing.T) {
...
+
+		pools := createMockPools()
+		pools.HeadersCalled = func() dataRetriever.HeadersPool {
+			sds := &mock.HeadersCacherStub{}
+			sds.GetHeaderByHashCalled = func(hash []byte) (data.HeaderHandler, error) {
+				if bytes.Equal(hash, prevHeaderHash1) {
+					return prevHeader, nil
+				}
+
+				return prevHeader, nil
```
[the remark this answers is not in the document]
Comment (ssd04) [reply process/sync/metablock_test.go:1895]: updated

File: process/sync/metablock_test.go [hunk 7]
```diff
@@ -1812,3 +1831,197 @@ func TestMetaBootstrap_SyncAccountsDBs(t *testing.T) {
...
```
[the remark this answers is not in the document]
Comment (ssd04) [reply process/sync/metablock_test.go:1937]: updated

File: process/sync/metablock_test.go [hunk 8]
```diff
@@ -1812,3 +1831,197 @@ func TestMetaBootstrap_SyncAccountsDBs(t *testing.T) {
...
+		}
+
+		header := &block.MetaBlock{
+			Nonce:    11,
+			PrevHash: prevHeaderHash1,
+		}
+
+		nextHeader := &block.MetaBlock{
+			Nonce:    12,
+			PrevHash: prevHeaderHash1,
```
[the remark this answers is not in the document]
Comment (ssd04) [reply process/sync/metablock_test.go:1974]: updated to set a different prev header hash

File: process/sync/baseSync.go [hunk 9]
```diff
@@ -491,6 +494,9 @@ func checkBaseBootstrapParameters(arguments ArgBaseBootstrapper) error {
 	if arguments.ProcessWaitTime < minimumProcessWaitTime {
 		return fmt.Errorf("%w, minimum is %v, provided is %v", process.ErrInvalidProcessWaitTime, minimumProcessWaitTime, arguments.ProcessWaitTime)
 	}
+	if check.IfNil(arguments.EnableEpochsHandler) {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply process/sync/baseSync.go:497]: added nil check

File: epochStart/bootstrap/epochStartMetaBlockProcessor.go [hunk 10]
```diff
@@ -26,14 +27,25 @@ const minNumConnectedPeers = 1
...
+	hasher              hashing.Hasher
+	enableEpochsHandler common.EnableEpochsHandler
+
+	mutReceivedMetaBlocks  sync.RWMutex
+	mapReceivedMetaBlocks  map[string]data.MetaHeaderHandler
+	mapMetaBlocksFromPeers map[string][]core.PeerID
+
+	// TODO: refactor to use a separate component for meta block sync handling
+	//	for epoch start metablock and epoch start confirmation block
+	mutReceivedConfMetaBlocks  sync.RWMutex
```
[the remark this answers is not in the document]
Comment (ssd04) [reply epochStart/bootstrap/epochStartMetaBlockProcessor.go]: removed this map

File: epochStart/bootstrap/epochStartMetaBlockProcessor.go [hunk 11]
```diff
@@ -136,22 +157,50 @@ func (e *epochStartMetaBlockProcessor) Save(data process.InterceptedData, fromCo
...
+		e.mutReceivedMetaBlocks.Lock()
+		e.mapReceivedMetaBlocks[string(mbHash)] = metaBlock
+		e.addToPeerList(string(mbHash), fromConnectedPeer)
+		e.mutReceivedMetaBlocks.Unlock()
+
 		return nil
 	}
 
-	mbHash := interceptedHdr.Hash()
+	if e.isEpochStartConfirmationBlock(metaBlock) {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply epochStart/bootstrap/epochStartMetaBlockProcessor.go]: updated conditions here

File: epochStart/bootstrap/epochStartMetaBlockProcessor.go [hunk 12]
```diff
@@ -136,22 +157,50 @@ func (e *epochStartMetaBlockProcessor) Save(data process.InterceptedData, fromCo
...
-	e.mutReceivedMetaBlocks.Lock()
-	e.mapReceivedMetaBlocks[string(mbHash)] = metaBlock
-	e.addToPeerList(string(mbHash), fromConnectedPeer)
-	e.mutReceivedMetaBlocks.Unlock()
+	log.Debug("received metablock is not of type epoch start", "error", epochStart.ErrNotEpochStartBlock)
 
 	return nil
 }
 
+func (e *epochStartMetaBlockProcessor) isEpochStartConfirmationBlock(metaBlock data.HeaderHandler) bool {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply epochStart/bootstrap/epochStartMetaBlockProcessor.go]: renamed

File: process/interceptors/processor/hdrInterceptorProcessor.go [hunk 13]
```diff
@@ -68,6 +80,13 @@ func (hip *HdrInterceptorProcessor) Save(data process.InterceptedData, _ core.Pe
 
 	hip.headers.AddHeader(interceptedHdr.Hash(), interceptedHdr.HeaderHandler())
 
+	if hip.enableEpochsHandler.IsFlagEnabledInEpoch(common.EquivalentMessagesFlag, interceptedHdr.HeaderHandler().GetEpoch()) {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply process/interceptors/processor/hdrInterceptorProcessor.go:83]: updated check

File: process/sync/baseSync.go [hunk 14]
```diff
@@ -715,6 +767,24 @@ func (boot *baseBootstrap) cleanNoncesSyncedWithErrorsBehindFinal() {
 	}
 }
 
+func (boot *baseBootstrap) cleanProofsBehindFinal(header data.HeaderHandler) {
+	if !boot.enableEpochsHandler.IsFlagEnabledInEpoch(common.EquivalentMessagesFlag, header.GetEpoch()) {
+		return
+	}
+
+	finalNonce := boot.forkDetector.GetHighestFinalBlockNonce()
```
[the remark this answers is not in the document]
Comment (ssd04) [reply process/sync/baseSync.go:775]: added todo comment for now

**Review by sstanculeanu — DISMISSED** [review]

**Review by AdoAdoAdo — DISMISSED** [review]

**Review by sstanculeanu — APPROVED** [review]
